### ISSUE TITLE
Libation hangs in Liberate

### ISSUE THREAD

**Issue by ZeroNamesLeft** [rmcrackan/Libation#158]
[Log202111_006.log](https://github.com/rmcrackan/Libation/files/7599011/Log202111_006.log)
![Screenshot 2021-11-24 Libation 172624](https://user-images.githubusercontent.com/93274218/143320930-5695f1c5-6889-4ee5-a82e-9b4d0c7a403d.png)

I am trying to backup what I believe is a podcast. I understand that might be the issue, but wanted to make sure. 

Screen shot is where the process hangs.

Thanks!

**Comment by rmcrackan** [comment]
Is [this](https://www.audible.com/pd/Chapter-3-Comparing-Reality-To-Its-Alternatives-Podcast/B09D8D4YMQ) the only book which gives you trouble? Or have others also?

**Comment by ZeroNamesLeft** [comment]
Yes, that is the only book.

**Comment by rmcrackan** [comment]
@Mbucari This occurs for this free podcast because in NetworkFileStream.BeginDownloading() , `response?.Headers?.GetValues("Accept-Ranges");` is null therefore this fails:  
`if (response.Headers.GetValues("Accept-Ranges").FirstOrDefault(r => r.EqualsInsensitive("bytes")) is null)`

With this commented out, the download works. What might I possibly break by removing this?

**Comment by Mbucari** [comment]
So if a download does not support HTTP ranges, it will not be resumeable. I'm on vacation and can't test, but I suspect if you remove that line you will break resumable downloads for files that don't support ranges. If a user interrupts the download of a non-resumable audio file, and then attempts to resume it, NetworkFileStream will reconnect to the HTTP endpoint and request the range from where it left off. If that file doesn't support ranges, I suspect it will serve up the file from the beginning. But since libation is appending the download to the end of the temporary file, you'll end up with a broken audio file. When I made NetworkFileStream, I assumed all audible audio files would accept HTTP ranges. Since we now know that's not the case, I think the easiest way to handle this is to add a property to NetworkFileStream that tracks whether the endpoint supports HTTP ranges. If it doesn't, then you need to restart the download from the beginning and not use any of the partially downloaded file from a previously interrupted download.

**Comment by Mbucari** [comment]
OK, so this turned out to be pretty easy. Just delete lines 196 and 197. These:

```C#
if (response.Headers.GetValues("Accept-Ranges").FirstOrDefault(r => r.EqualsInsensitive("bytes")) is null)
                throw new WebException($"Server at {Uri.Host} does not support Http ranges");

```

It turns out that check is redundant with StatusCode check above.  If the response status is PartialContent, then it supports HTTP ranges regardless of what the server's headers claim.  This doesn't solve the case where Audible serves up a file that doesn't support ranges, but we have not found any of those cases yet so I see no reason to preemptively add that capability to NetworkFileStream.

**Comment by rmcrackan** [comment]
@Mbucari Thanks. This was my original idea too but removing a safe guard felt unwise without further insight. I see your point about the other check. Upgrade will go out soon.

**Comment by rmcrackan** [comment]
@ZeroNamesLeft Thanks for helping guide the debug efforts. Fixed in [v6.6.2](https://github.com/rmcrackan/Libation/releases/latest).

@Mbucari Thanks for the advice
